### ISSUE TITLE
bitsets might need to be aware of holes

### ISSUE THREAD

**Issue by staltz** [ssbc/jitdb#223]
I stumbled upon a big headache kind of problem. Pagination seems to give wrong results if there are deleted records (and so far this has nothing to do with compaction, you can assume we're not running any compaction).

Example below:

```js
  const content0 = { type: 'post', text: 'Testing 0' }
  const content1 = { type: 'post', text: 'Testing 1' }
  const content2 = { type: 'post', text: 'Testing 2' }
  const content3 = { type: 'post', text: 'Testing 3' }
  const content4 = { type: 'post', text: 'Testing 4' }
  const content5 = { type: 'post', text: 'Testing 5' }
  const content6 = { type: 'post', text: 'Testing 6' }
  const content7 = { type: 'post', text: 'Testing 7' }
  let state = validate.initial()
  state = validate.appendNew(state, null, keys, content0, Date.now())
  state = validate.appendNew(state, null, keys, content1, Date.now() + 1)
  state = validate.appendNew(state, null, keys, content2, Date.now() + 2)
  state = validate.appendNew(state, null, keys, content3, Date.now() + 3)
  state = validate.appendNew(state, null, keys, content4, Date.now() + 4)
  state = validate.appendNew(state, null, keys, content5, Date.now() + 5)
  state = validate.appendNew(state, null, keys, content6, Date.now() + 6)
  state = validate.appendNew(state, null, keys, content7, Date.now() + 7)

  const offset0 = (await addMsgPromise(state.queue[0].value, log)).offset
  const offset1 = (await addMsgPromise(state.queue[1].value, log)).offset
  const offset2 = (await addMsgPromise(state.queue[2].value, log)).offset
  const offset3 = (await addMsgPromise(state.queue[3].value, log)).offset
  const offset4 = (await addMsgPromise(state.queue[4].value, log)).offset
  const offset5 = (await addMsgPromise(state.queue[5].value, log)).offset
  const offset6 = (await addMsgPromise(state.queue[6].value, log)).offset
  const offset7 = (await addMsgPromise(state.queue[7].value, log)).offset

  // IMPORTANT! Causes the bitvector index to have offset1 and offset2
  await pify(jitdb.prepare)(slowEqual('value.content.type', 'post'))

  await pify(log.del)(offset1)
  t.pass('delete msg 1')
  await pify(log.del)(offset2)
  t.pass('delete msg 2')
  await pify(log.onDeletesFlushed)()

  const source = query(
    fromDB(jitdb),
    where(slowEqual('value.content.type', 'post')),
    paginate(3),
    toPullStream()
  )

  let msgs = await pify(source)(null)
  console.log(msgs.map(msg => msg.value.content.text));

  msgs = await pify(source)(null)
  console.log(msgs.map(msg => msg.value.content.text));

  msgs = await pify(source)(null)
  console.log(msgs.map(msg => msg.value.content.text));
```

## Expected

`[ 'Testing 0', 'Testing 3', 'Testing 4' ]`

`[ 'Testing 5', 'Testing 6', 'Testing 7' ]`

`[]`

## Actual

`[ 'Testing 0' ]`

`[ 'Testing 3' ]`

`[ 'Testing 3', 'Testing 4' ]`

## Discussion

I have an idea that would work but requires some effort. I'd be happy to find a decent solution that requires less effort.

My idea is that we could keep a bitvector index called `deleted.index` which is just a bitset of the deleted records. This way, when producing a bitset, we can always `AND(NOT(deleted, x))` before passing it to `getMessagesFromBitsetSlice`. To achieve this, we might have to "listen" to delete events that AAOL could emit.

cc @arj03

**Comment by arj03** [comment]
Hmm. There is def. a problem here and it is not only related to paginate. A query with a limit and deleted records will also fail to produce the correct results. 

I tried a simple way to fix this by taking the [filter path](https://github.com/ssbc/jitdb/blob/398d6f33f3d1cbbe3e4ef22d1cce35b67fdd9d43/index.js#L1217) and extending that to be more general, by properly handling deletes and always running that path. Sadly there is more overhead in the benchmarks that I would like, so I think what you are thinking about with a `deleted.index` index is probably the right approach anyway. I wonder if we could do something like: in `processResults` if we encounter deleted records we don't know about yet, then update the deleted.index and rerun the whole thing (getMessagesFromBitsetSlice). Not sure if that is what you meant with listening to delete events.

**Comment by staltz** [comment]
> I tried a simple way to fix this by taking the [filter path](https://github.com/ssbc/jitdb/blob/398d6f33f3d1cbbe3e4ef22d1cce35b67fdd9d43/index.js#L1217) and extending that to be more general, by properly handling deletes and always running that path. Sadly there is more overhead in the benchmarks that I would like

Oh, I didn't consider that path, is this something you wrote in a branch somewhere?

> Not sure if that is what you meant with listening to delete events.

I meant that we would add an API to AAOL: `log.onDelete` which you can add a listener function and it would call you whenever a delete happens. But I'll first try your idea, which is to rerun the whole thing if we bump into new deletes.

**Comment by staltz** [comment]
I started working on `deleted.index` (actually, because we're always going to be doing an `AND(NOT(` of that, it makes sense to store the inverse, `nonDeleted.index`, which I ended up calling `present.index`) and how `getMessagesFromBitsetSlice` will rerun, but I ran into a problem: if you call `getMessagesFromBitsetSlice` with `onlyOffsets = true`, then the function will have no way of knowing whether the record is deleted or not.

So what I'll try next is to implement a simple `log.onDeleted` and see if I can make progress with that. I'm thinking that the bitset given to `getMessagesFromBitsetSlice` should already come filtered with deletes removed. It simplifies the code a lot.

**Comment by arj03** [comment]
>  is this something you wrote in a branch somewhere?

Just for you to check

https://github.com/ssbc/jitdb/compare/master...failed-filter-deleted

It passes the test you wrote, but fails some of the normal tests besides being slower ;-)

> which I ended up calling present.index

:+1: 

> if you call getMessagesFromBitsetSlice with onlyOffsets = true, then the function will have no way of knowing whether the record is deleted or not

true I also thought about that, but I think we could sort of expect that it is the callers responsibility to filter out deletes. 

> So what I'll try next is to implement a simple log.onDeleted and see if I can make progress with that.

I'm a bit skeptical about that approach as it seems a bit brittle. Like what happens if you delete but don't save that in the index in jitdb? You might need some sort of log of deletes (in AAOL and this needs to be atomic) for this to work and then jitdb can sync with that on query.

**Comment by staltz** [comment]
We fixed this in #225 by allowing pages smaller than the requested page size, as long as there are no duplicate items (in any pages).
